### ISSUE TITLE
rtsp/rtmp decode Error occurred for picture xxx

### ISSUE THREAD

**Issue by niaoyu** [NVIDIA/VideoProcessingFramework#39]
Hi, @rarzumanyan 
```bash
Decode Error occurred for picture 800
```
After a careful testing, I found the error would be most related to the rtsp url itself after 1/2 minutes.
So is there any solution to ignore the that? Just like opencv which also use ffmpeg as backend will run smoothly for a long time.

Just mentioned here, ~~after the vpf error shows in the output, the decoding process would be stucking.~~(actually the empty output of DecodeSingleSurface stops the decoding thread) What I could do is just restart the decoding thread I guess. 

Any good solutions?
Best wishes.

**Comment by mchl0203** [comment]
Hi @niaoyu :
I guess the cause of this problem is that the cache is full. I try to set the decoding time is less than the transmission speed(1 / FPS) of rtsp video stream, this error will no longer appear.

**Comment by rarzumanyan** [comment]
Hi @niaoyu 
If I launch decode sample on public RTSP stream without any additional ffmpeg parameters or whatsoever
```bash
python SampleDecode.py 0 rtsp://freja.hiof.no:1935/rtplive/definst/hessdalen03.stream output.nv12
```

I often see the decode error message in console output:
```bash
[rtsp @ 0000024e8d3fe480] RTP: missed 70 packets
Decode Error occurred for picture 37
[rtsp @ 0000024e8d3fe480] max delay reached. need to consume packet
[rtsp @ 0000024e8d3fe480] RTP: missed 2 packets
Decode Error occurred for picture 203
[rtsp @ 0000024e8d3fe480] max delay reached. need to consume packet
[rtsp @ 0000024e8d3fe480] RTP: missed 5 packets
Decode Error occurred for picture 534
[rtsp @ 0000024e8d3fe480] max delay reached. need to consume packet
[rtsp @ 0000024e8d3fe480] RTP: missed 8 packets
Decode Error occurred for picture 1219
[rtsp @ 0000024e8d3fe480] max delay reached. need to consume packet
[rtsp @ 0000024e8d3fe480] RTP: missed 1 packets
Decode Error occurred for picture 1640
[rtsp @ 0000024e8d3fe480] max delay reached. need to consume packet
[rtsp @ 0000024e8d3fe480] RTP: missed 11 packets
```

But it doesn't crashes or stops application as it kepps going on after a frame decoding failure.
The same thing happens to `SampleDecodeMultiThread.py`, it also keeps going on after a decode failure.

IMO this is application design question. If you want to work with the network input, your application has to be robust to connectivity losses or packets corruption.

Some frames have corrupted regions due to packet loss or timeout but nothing fatal happens:

![output_1822_rgb](https://user-images.githubusercontent.com/45193474/81280964-ee274580-9061-11ea-8b65-d17b5de975f4.jpg)

In the end of the day it's a decoding sample and not a full-blown RTSP detection application.

**Comment by niaoyu** [comment]
> Hi @niaoyu :
> I guess the cause of this problem is that the cache is full. I try to set the decoding cost time is less than the transmission speed(1 / FPS) of rtsp video stream, this error will ~no longer appear~ be reduced.....

Thanks for your advice.I just test the cache issue and confirm you are right.

1) change the torch gpu tensor to a numpy array and put it to the queue.And the error still happens.
```python
        is_detected=True
        detect_output = None

        image_rgb = surface_tensor.cpu().detach().numpy()
        image_rgb = np.transpose(image_rgb, (1, 2, 0))

        # image_rgb = np.random.rand(3,224,224)
        self.stream_queue.put(
            {
                'image_rgb': image_rgb,
                'is_detected': is_detected,
                'detect_outputs': detect_output
            }
        )
```

2) just use a numpy array, and print the surface tensor shape, it would be fine.
```python

        is_detected=True
        detect_output = None

        # image_rgb = surface_tensor.cpu().detach().numpy()
        # image_rgb = np.transpose(image_rgb, (1, 2, 0))


        print(self.surface_tensor.shape)
        image_rgb = np.random.rand(3,224,224)
        self.stream_queue.put(
            {
                'image_rgb': image_rgb,
                'is_detected': is_detected,
                'detect_outputs': detect_output
            }
        )
```

**Comment by niaoyu** [comment]
@rarzumanyan 

```python
But it doesn't crashes or stops application as it kepps going on after a frame decoding failure.
```

Errrr.....for my case, if the output shown as 'Decode Error occurred for picture 393', there would be no more decoding tensors.

**Comment by mchl0203** [comment]
Hi @niaoyu :
> change the torch gpu tensor to a numpy array and put it to the queue.And the error still happens.

yeah, decoding into tensor will indeed cause problems. 
And I tested multiple video streams(to tensor), 1 GPU can only decode 2 video streams. Decoding More than 2 channels can only decode 2 channels, other video streams cannot be decoded and the process stops at the following line of code:
```
surface_tensor = pnvc.makefromDevicePtrUint8(surf_plane.GpuMem(), surf_plane.Width(), surf_plane.Height(),
                                             surf_plane.Pitch(), surf_plane.ElemSize())
```
I will test it carefully, and will update it if there are new change...

**Comment by rarzumanyan** [comment]
@niaoyu 
>Errrr.....for my case, if the output shown as 'Decode Error occurred for picture 393', there would be no more decoding tensors.

Sorry, I don't get it.
Do you run into situation when after first decode error all consequent decode calls also return errors?

**Comment by niaoyu** [comment]
> @niaoyu
> 
> > Errrr.....for my case, if the output shown as 'Decode Error occurred for picture 393', there would be no more decoding tensors.
> 
> Sorry, I don't get it.
> Do you run into situation when after first decode error all consequent decode calls also return errors?

@rarzumanyan @mchl0203 
Sorry for my mistake. 

After finish my careful test, I figured out  the decoding process is still running regularly even if the 'Decode Error occurred for picture xxx' happens. Also the problem is not related to the numpy array output or torch tensor output.

The  real problem is the output of **DecodeSingleSurface empty** error. 
Just like decoding rtmp/rtsp urls here,  I will always meet the error 'rawSurfaceNV12 error', and the decoding process is down.

```python
                rawSurfaceNV12 = self.nv_dec.DecodeSingleSurface()
                if (rawSurfaceNV12.Empty()):
                    print('rawSurfaceNV12 error')

```

Is there anything I could do to solve the issue?

**Comment by rarzumanyan** [comment]
>The real problem is the output of DecodeSingleSurface empty error. If decoding more than two rtmp/rtsp urls, the Empty() error will happen and the decoding process is down

So VPF can't decode particular video frame due to network connectivity issues (timeout, packet corruption etc.). There are two ways out: you can either break the decoding loop (as it's done in sample) or skip the loop iteration and continue to next frame:
```python
rawSurface = self.nvDec.DecodeSingleSurface()
if (rawSurface.Empty()):
    print('Can't decode this video frame')
    continue
```
You will not have anything exported to Pytorch tensor simply because there's nothing to export, the video frame is empty.

**Comment by niaoyu** [comment]
> > The real problem is the output of DecodeSingleSurface empty error. If decoding more than two rtmp/rtsp urls, the Empty() error will happen and the decoding process is down
> 
> So VPF can't decode particular video frame due to network connectivity issues (timeout, packet corruption etc.). There are two ways out: you can either break the decoding loop (as it's done in sample) or skip the loop iteration and continue to next frame:
> 
> ```python
> rawSurface = self.nvDec.DecodeSingleSurface()
> if (rawSurface.Empty()):
>     print('Can't decode this video frame')
>     continue
> ```
> 
> You will not have anything exported to Pytorch tensor simply because there's nothing to export, the video frame is empty.

I have tried this version which use "continue "rather than "break".

But it will always be "Can't decode this video frame And continue "for ever.... No more available frames could be achieved

**Comment by rarzumanyan** [comment]
>What i could do is just restarting the decoding thread.

Why not siply re-create decoder instance?
```python
rawSurface = self.nvDec.DecodeSingleSurface()
if (rawSurface.Empty()):
    print('Can't decode this video frame')
    self.nvDec = PyNvDecoder(...)
    continue
```

>So maybe besides connectivity issues (timeout, packet corruption etc.),there are some cuda issues in vpf?

It's really difficult to trace how broken compressed frame may influence internal state of HW decoder. Besides, if e. g. null input is given to HW decoder it will treat it as flush signal and will decode all the remaining frames in the queue etc.

**Comment by mchl0203** [comment]
Hi @niaoyu :
In my envs: Video_Codec_SDK_9.0.20, CUDA 10.1, torch 1.4. I have the same problem as you. And when I `import PytorchNvCodec as pnvc`, even decoding into numpy, it can only decode 2 video streams.
However, In envs: Video_Codec_SDK_9.1.23, CUDA 10.2, torch 1.4. this problem disappeared. You can try it.

**Comment by niaoyu** [comment]
> ```python
> self.nvDec = PyNvDecoder(...)
> ```

Thanks, I just adopt your advice, the rtmp urls run smoothly!

The rtsp urls may have too many issues which I will not care about.

Best wishes!

**Comment by niaoyu** [comment]
> @niaoyu
> 
> > Errrr.....for my case, if the output shown as 'Decode Error occurred for picture 393', there would be no more decoding tensors.
> 
> Sorry, I don't get it.
> Do you run into situation when after first decode error all consequent decode calls also return errors?

Actually, after the firt decode error appears like 'Decode Error occurred for picture...'. There would be no more outputs tensors printing. Seems like to 

> Hi @niaoyu :
> In my envs: Video_Codec_SDK_9.0.20, CUDA 10.1, torch 1.4. I have the same problem as you. And when I `import PytorchNvCodec as pnvc`, even decoding into numpy, it can only decode 2 video streams.
> However, In envs: Video_Codec_SDK_9.1.23, CUDA 10.2, torch 1.4. this problem disappeared. You can try it.

Thanks a lot!
My env:  Video_Codec_SDK_9.1.23, CUDA 10.1, torch 1.5, ffmpeg version N-97463-g54f100272a

And after I choosed some stable rtmp urls to test,  the only problems whill be the Empty output of DecodeSingleSurface(). Now what I could do is just re-initial 
```python
self.nv_dec = nvc.PyNvDecoder()
```

Have you just met this kind of problem?
